### ISSUE TITLE
Validating environment variables with Astro Server-side Rendering

### ISSUE THREAD

**Issue by intagaming** [t3-oss/t3-env#60]
*Note: The following is kind of like a recipe page awaiting its debut in the t3-env docs. Please consider this a proposal before I (or someone else) do so.*

When Astro is built with the default Static Site Generation (SSG), the validation will happen on each static route generation at the place of `env` usage. Technically this is "Validation on build".

However, with Astro Server-side Rendering (SSR), validation takes a different "route" (no pun intended). If we just switch from SSG to SSR with `npm astro add node` for example, immediately we will see that no validation is run when running `npm run build`. Let's force it by importing the validation file in `astro.config.ts`:

```ts
import "./src/env-validate";

// ... or if delaying the validation for later on in the file:
await import("./src/env-validate");
```

Now `npm run build` will tell us that our environment variables are nowhere to be found in `import.meta.env`. How comes?

With Astro SSR, you have 1 chance at running the validation, which is the moment when `astro.config.ts` is evaluated at the beginning of the build. Unfortunately `import.meta.env` at this "config reading time" is not populated yet, so you have to fallback to `process.env` or reading the `.env` files manually with Vite's `loadEnv` function. Thus putting `import.meta.env` into `runtimeEnv` doesn't work anymore with Astro SSR.

When working with Astro SSR, because there are the build-time and the runtime, there exists different types of environment variables:

## Server-side build-time variables
If we can't use `import.meta.env` in the `env-validate.ts` file, how can Vite replace a variable at build time *without prefixing it with PUBLIC_* (which is Client-side build-time variables)? The answer is using Vite's [`define` function](https://vitejs.dev/config/shared-options.html#define), which will have Vite replace something in the source code with something else. Exactly what we asked for.

For example, I have a `BUILDVERSION` environment variable that is present when running `npm run build`. This variable must be etched into the source code at build time, and must only be present on the server-side only (so no `PUBLIC_` prefixed variables - which is _Build-time client variables_). I have this config in the `astro.config.ts` file:

```ts
  vite: {
    define: {
      "process.env.BUILDVERSION": JSON.stringify(process.env.BUILDVERSION),
    },
  },
```

Now every single `process.env.BUILDVERSION` in my source code will be replaced with whatever `process.env.BUILDVERSION` is at build-time. You can choose something else like `__BUILD_VERSION__` but you have an additional hassle of dealing with TypeScript, and I'm in a hustle right now.

In `runtimeEnv` I now have this:

```ts
const isServer = typeof window === "undefined";

// ... some more code
  runtimeEnv: {
    BUILDVERSION: isServer ? process.env.BUILDVERSION : undefined,

```

This Server-side build-time variable is:
- `undefined` on client scripts.
- Replaced by Vite's `define` at build time.

## Server-side runtime variables
You don't want to validate server-side runtime variables at build time, because you don't have any yet. So you have to skip runtime variables. Here's the trick. Please look at the code and see for yourself.

```ts
const isBuilding = isServer && process.env.BUILDING === "true";

export const serverValidation = {
  build: {
    // Put in here any server-side variable that is hardcoded on build, i.e.
    // defined with Vite config's `define` option.
    BUILDVERSION: z.string().nonempty(),
  },
  runtime: {
    // Put in here any server-side variable that is read at runtime, i.e. being
    // read from `process.env`.
    NODE_ENV: z.enum(["development", "test", "production"]),
    API_HOST: z.string().nonempty(),
    APP_DOMAIN: z.string().nonempty(),
  },
};

export const env = createEnv({
  server: (isBuilding
    ? serverValidation.build
    : {
        ...serverValidation.build,
        ...serverValidation.runtime,
      }) as typeof serverValidation.build & typeof serverValidation.runtime,
  runtimeEnv: {
    NODE_ENV: isServer ? process.env.NODE_ENV : undefined,
    // Just skip the devEnv function for now; explanation below.
    API_HOST: isServer ? process.env.API_HOST ?? devEnv("API_HOST") : undefined,
    APP_DOMAIN: isServer
      ? process.env.APP_DOMAIN ?? devEnv("APP_DOMAIN")
      : undefined,

```

When building, set `BUILDING=true` and it will skip runtime vars validation.

These server-side runtime variables are:
- `undefined` on client scripts.
- Read from `process.env` on server-side.

These variables are free to change at anytime without rebuilding the project. Just change the env var and restart the server, hence the "runtime" in the name.

Now what is the `devEnv()`? The problem `devEnv()` is solving is as follows. The above `runtimeEnv` without the `devEnv()` works just fine when you're running `npm run build` as well as starting the Astro server, but it might not work in development, i.e. `npm run dev`. The problem is that Astro's development mode is not building so it doesn't have `BUILDING=true` -> it is also validating runtime variables, thus in `runtimeEnv` it is accessing `process.env`, which is your environment, but what if your variables are in a `.env` file? `devEnv` will read from the `.env` file as a fallback to `process.env`.

```ts
// src/env-validate.ts
const devEnv = (key: string) => fileEnv[key] || import.meta.env[key];
```

```ts
// astro.config.ts
import fileEnv from "./src/fileEnv";

// Assign .env to the fileEnv object. See fileEnv.ts for more details.
Object.assign(fileEnv, loadEnv(import.meta.env.MODE, process.cwd(), ""));
// Validate the environment variables
await import("./src/env-validate");
```

```ts
// src/fileEnv.ts
// (Server-side only) This file contains an exported object that will be
// assigned the content of the .env files (if any) by the astro.config.ts file.
//
// The purpose is to use variables in the .env file in the
// ./src/env-validate.ts file without that file having to use Vite's `loadEnv`
// function, causing the `fsevents` module to be bundled in the client-side
// bundle.
const fileEnv: Record<string, string> = {};

export default fileEnv;
```

So why there is `import.meta.env` in the `devEnv()`? That's because after starting the development server, `fileEnv` actually goes empty, and everything in your `.env` file goes into `import.meta.env` (yes, not to be mistaken to `process.env`). Surprise? I actually have no idea at the moment how Astro does that. So when your server-side code calls `env.`, it doesn't find anything in `process.env`, and there's nothing in `fileEnv`, but it is in `import.meta.env` so we'll use that at development for server-side runtime vars.

## Client-side build-time variables
First off, there is no Client-side runtime variable, because all of the Client-side bundles are generated at build time, not on the fly - which might as well be server-side runtime variables passed down to the client <- bad idea.

```ts
  client: {
    PUBLIC_EXAMPLE_HARDCODED_CLIENT_VAR: z.string().nonempty(),
  },
  runtimeEnv: {
    PUBLIC_EXAMPLE_HARDCODED_CLIENT_VAR: isBuilding
      ? process.env.PUBLIC_EXAMPLE_HARDCODED_CLIENT_VAR // This will go to `import.meta.env` after build.
      : import.meta.env.PUBLIC_EXAMPLE_HARDCODED_CLIENT_VAR ??
        fileEnv.PUBLIC_EXAMPLE_HARDCODED_CLIENT_VAR,

```

Client-side build-time variables are:
- Accessible both on server-side and client-side.
- `import.meta.env` are replaced at build time.
- Not overridable by the environment on the server-side (because it's suppose to be hardcoded at build).

The `fileEnv` above will provide the variable for validation on development server start (because `import.meta.env` at that time is not populated), after that it will be injected into `import.meta.env` on runtime.

---
So that's 3 kinds of environment variables in Astro SSR. It got spaghettied pretty fast, but that's the best I can come up with.

I might provide a sandbox later.

**Comment by alexanderniebuhr** [comment]
Thank you for the write-up. I had to read it 3 times, before I understood it, and most things work. However in Cloudflare Workers, I do have issues setting up server-side runtime vars. The `process.env` is available and provides the vars, however it seems that in the build-step process-env will be replaced by `{}`. Have you made it work with Cloudflare and Astro?

```
CF_PAGES_URL: isServer ? {}.CF_PAGES_URL : void 0,
```

**Comment by intagaming** [comment]
> Thank you for the write-up. I had to read it 3 times, before I understood it, and most things work. However in Cloudflare Workers, I do have issues setting up server-side runtime vars. The `process.env` is available and provides the vars, however it seems that in the build-step `process.env` will be replaced by `{}`, in the `createEnv()`. Have you made it work with Cloudflare and Astro?
> 
> 
> 
> ```
> 
> CF_PAGES_URL: isServer ? {}.CF_PAGES_URL : void 0,
> 
> ```

Weird, since the Astro build step has nothing to do with Cloudflare Workers.

Can you run the Astro build script locally and confirm that the `process.env` is indeed the replaced by Astro? Also can you confirm that Cloudflare's build step doesn't do anything special other than running the build script?

**Comment by alexanderniebuhr** [comment]
My bad explaining. I'm running the local build step `astro build`, but with the `@astrojs/cloudflare` adapter. And the adapter does the replacement.
However on Cloudflare Pages the env variables are not available on start, only after the request is received. That means that validation will fail on start of the runtime.
So I made everything work, except server-side runtime vars. Not sure what it takes to make that work with Cloudflare Pages, the `@astrojs/node` is much easier.

**Comment by intagaming** [comment]
@alexanderniebuhr If I'm not mistaken, `@astrojs/cloudflare` is using `import.meta.env` as the place for runtime environment variable, yes? Then using `process.env` in `runtimeEnvStrict` here is actually incorrect. `process.env` should be `import.meta.env` (for server-side runtime env vars that is). Can you try that and report back to see if that's the culprit?

Found this in `_worker.js` that leads me to the conclusion: https://github.com/withastro/astro/blob/main/packages/integrations/cloudflare/src/util.ts

**Comment by alexanderniebuhr** [comment]
> @astrojs/cloudflare is using import.meta.env as the place for runtime environment variable

Kinda, but the runtime variables are set in Cloudflare after the build is finished. So really astro has nothing to do with it.
And in the Cloudflare Pages runtime, both are available. The variables you set in the dashboard or wrangler cli, are in `process.env`, as far as I debugged it. So we don't know server-side runtime variables at build-time.

If replaced with `import.meta.env`, it still does not work. Even with custom `loadEnv`. It is also used in the old code already by `devEnv`. 

If I replace the `{}`, with `process.env` in the build output it seems to work, I've asked in the Astro discord if, and update this accordingly.

**Comment by alexanderniebuhr** [comment]
> Cloudflare's runtime environment variables are not in `process.env`

Actually they are accessible by `process.env`, so either that is an **alias** or the documentation is not quite right.

See this example (https://f048ceb6.test-ef9.pages.dev/)

**Comment by intagaming** [comment]
> > pass getRuntime().env to t3-env's runtimeEnv somehow for each request
> 
> That is correct, however not sure if that is possible & if that would work on initial start before a request is send.
> 
> > Cloudflare's runtime environment variables are not in `process.env`
> 
> Actually they are accessible by `process.env`, so either that is an **alias** or the documentation is for workers only and not pages.
> 
> See this example, sadly `import.meta.env` get's replaced in the src code.. (https://f048ceb6.test-ef9.pages.dev/)

Yes, all of that is correct. Though it is certainly possible. You are given a little space on a Worker to perform tasks, which means each of your request has a different environment, which means you have to run the validation **per request** i.e. **at the start of each individual request**, not at initial start. There's no initial start in Cloudflare Workers, the workers are not your app's, it is shared between possibly thousands of apps.

Move `createEnv` into a function, say `createT3Env`. At the beginning of your Astro server code, instead of importing `env`, you call `createT3Env(getRuntime(Astro.request).env)`; use that input as the replacement for the usual `process.env`.

**Comment by alexanderniebuhr** [comment]
@intagaming came up with the following, looks kinda hacky! And I also feel not so good that the createEnv is called all the time, because of the wrapped function. But seems to work.

So for me it is:

`astro.config.ts`
```ts
import fileEnv from "./src/fileEnv"
import { createT3Env } from "./src/t3-env"
// Assign .env to the fileEnv object. See fileEnv.ts for more details.
Object.assign(fileEnv, loadEnv(import.meta.env.MODE, process.cwd(), ""))
// Validate the environment variables
createT3Env()
```
`src/fileEnv.ts`
```ts
// src/fileEnv.ts
// (Server-side only) This file contains an exported object that will be
// assigned the content of the .env files (if any) by the astro.config.ts file.
//
// The purpose is to use variables in the .env file in the
// ./src/env-validate.ts file without that file having to use Vite's `loadEnv`
// function, causing the `fsevents` module to be bundled in the client-side
// bundle.
const fileEnv: Record<string, string> = {}

export default fileEnv
```
`src/t3-env.ts`
```ts
import { createEnv } from "@t3-oss/env-core"
import { z } from "zod"
import fileEnv from "./fileEnv"

const isServer = typeof window === "undefined"
const isBuilding = isServer && process.env.BUILDING === "true"
const devEnv = (key: string) => fileEnv[key] || import.meta.env[key]
const serverValidation = {
	build: {
		// Put in here any server-side variable that is hardcoded on build, i.e.
		// defined with Vite config's `define` option.
		TIME: z.string().nonempty(),
	},
	runtime: {
		// Put in here any server-side variable that is read at runtime, i.e. being
		// read from `process.env`.
		CF_PAGES_URL: z.string().nonempty(),
	},
}
// rome-ignore lint/suspicious/noExplicitAny: <explanation>
export const env = (runtimeEnv: any = process.env) =>
	createEnv({
		server: (isBuilding
			? serverValidation.build
			: {
					...serverValidation.build,
					...serverValidation.runtime,
			  }) as typeof serverValidation.build & typeof serverValidation.runtime,
		clientPrefix: "PUBLIC_",
		client: {
			PUBLIC_API_URL: z.string(),
		},
		// We can't use vite's import meta env here, because it's not loaded yet
		runtimeEnvStrict: {
			TIME: isServer ? process.env.TIME : undefined,
			CF_PAGES_URL: isServer ? runtimeEnv.CF_PAGES_URL ?? devEnv("CF_PAGES_URL") : undefined,
			PUBLIC_API_URL: isBuilding
				? process.env.PUBLIC_API_URL
				: import.meta.env.PUBLIC_API_URL ?? fileEnv.PUBLIC_API_URL,
		},
		skipValidation:
			!!process.env.SKIP_ENV_VALIDATION &&
			process.env.SKIP_ENV_VALIDATION !== "false" &&
			process.env.SKIP_ENV_VALIDATION !== "0",
	})

export function createT3Env(workerEnv?: unknown) {
	return env(workerEnv)
}
```
`src/pages/index.astro` or `src/scripts/client.ts`
```ts
import { createT3Env } from "src/t3-env"
const env = createT3Env()
```

**Comment by alexanderniebuhr** [comment]
> So for me to make it work with Cloudflare Pages SSR it is, the changes here!

**:warning: This will leak variables inside client islands, if env is imported inside them.**

I personally thing Astro support is far from ideal, IMO following is still missing (to be edited):

- Throw if variables are leaked as props
- Throw if variables are leaked inside an object
- Optimise client-side checks. How to validate on islands?

**Comment by intagaming** [comment]
I updated the original issue with my new setup. There were some misunderstandings of the Vite build step, so I documented the correct behavior in the code.

As for @alexanderniebuhr's solution:
- Beware that the solution is supposed to be a solution for Cloudflare Pages. Calling `createRuntimeEnv` for each file, or function, is wasteful **if** the argument doesn't change. `process.env` in the Node runtime doesn't change, but `getRuntime(Astro.request)?.env` in Cloudflare Pages does. Applying the solution of `createRuntimeEnv` for Node runtime is overkill and will lead to wasteful CPU cycles.

- In my solution, because I need to `loadEnv` before I import `t3-env.ts`, I have to `loadEnv` into an object somewhere so `t3-env.ts` can read from there. Definitely not `astro.config.ts` because loading `astro.config.ts` on the client-side doesn't make sense. Hence the `fileEnv.ts` instead of passing `loadEnv` into `createRuntimeEnv`.

- If you don't include `as typeof serverValidation.build & typeof serverValidation.runtime`, you won't have TypeScript completion for the `.runtime` object.

**Comment by alexanderniebuhr** [comment]
Agree! There are different solutions for different adapters, and your new one looks great for Node runtimes. We should maybe consider splitting this issue then :)
